internal/pathtree/tree.go

```diff
@@ -0,0 +1,148 @@
+package pathtree
+
+import (
+	"strings"
+
+	"github.com/serverless/event-gateway/functions"
+)
+
+// Node is a data structure, inspired by prefix tree, used for routing HTTP requests in the Event Gateway. It's used for creating tree structure
+// of segments in HTTP paths. Each segments is stored in separate node.
+type Node struct {
+	segment     string
+	children    map[string]*Node
+	functionID  *functions.FunctionID
+	parameter   string
+	isParameter bool
+}
+
+// NewNode creates new Node.
+func NewNode() *Node {
+	return &Node{
+		children: map[string]*Node{},
+	}
+}
+
+// Params is array for URL parameter
+type Params map[string]string
+
+// AddRoute adds route to the tree. This function will panic in case of adding conflicting parameterized paths.
+func (n *Node) AddRoute(route string, functionID functions.FunctionID) {
+	if route == "/" {
+		n.functionID = &functionID
+		return
+	}
+
+	segments := toSegments(route)
+	currentNode := n
+
+	for i, segment := range segments {
+		// look for static route
+		child, exists := currentNode.children[segment]
+		if !exists {
+			// look for param
+			child, exists = first(currentNode.children)
+			if !exists || !child.isParameter {
+				child = NewNode()
+				child.segment = segment
+				currentNode.children[segment] = child
+			}
+			if child.segment != segment {
+				panic("route " + route + " has a conflicting segment with existing route")
```

Comment:
@mthenw what do you think about also returning info about the conflicting route here. This would improve the UX a lot since the user gets an clear indication what to look at next.

Reply:
hmm, yeah, good suggestion. I'm not sure if I'm able to get that info here.

Reply:
I'll look into that.